modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/requestcontroller/filter/IdempotentRequestHandler.java

```diff
@@ -0,0 +1,142 @@
+ */
+package org.nuxeo.ecm.platform.web.common.requestcontroller.filter;
+
+import java.io.IOException;
+
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.kv.KeyValueService;
+import org.nuxeo.runtime.kv.KeyValueStore;
+
+/**
+ * Handles idempotency key in POST requests.
+ * <p>
+ * If {@link #HEADER_KEY} is found in the request header, will intercept request handling to:
+ * <li>mark the request as being processed
+ * <li>capture the response when request was processed without any error and store it
+ * <li>return the stored response if a subsequent request with the same key is processed again
+ * <li>return a conflict response if a request with the same key is processed while the first request is still in
+ * progress.
+ *
+ * @since 11.5
+ */
+public class IdempotentRequestHandler {
+
+    private static final Logger log = LogManager.getLogger(IdempotentRequestHandler.class);
+
+    /** @since 11.5 */
+    public static final String HEADER_KEY = "NuxeoIdempotencyKey";
+
+    /** @since 11.5 */
+    public static final String STORE_PROPERTY = "idempotency_keyvaluestore";
+
+    /** @since 11.5 */
+    public static final String TTL_SECONDS_PROPERTY = "idempotency_ttl_seconds";
+
+    /** @since 11.5 */
+    public static final String INPROGRESS_MARKER = "IDEMPOTENCY_INPROGRESS_MARKER";
+
+    /** @since 11.5 */
+    public static final String STATUS_PREFIX = "s_";
+
+    /** @since 11.5 */
+    public static final String CONTENT_PREFIX = "c_";
+
+    protected final String store;
+
+    protected final int ttl;
+
+    public IdempotentRequestHandler(FilterConfig filterConfig) {
+        // default values
+        int ttl = 330;
+        String store = "requestcontroller";
+        // retrieve potential configuration from filter
+        String ttlConfig = filterConfig.getInitParameter(IdempotentRequestHandler.TTL_SECONDS_PROPERTY);
+        if (ttlConfig != null) {
+            try {
+                ttl = Integer.valueOf(ttlConfig);
+            } catch (NumberFormatException e) {
+                log.error(e, e);
+            }
+        }
+        String storeConfig = filterConfig.getInitParameter(IdempotentRequestHandler.STORE_PROPERTY);
+        if (storeConfig != null) {
+            store = storeConfig;
+        }
+        this.store = store;
+        this.ttl = ttl;
+    }
+
+    public IdempotentRequestHandler(String storename, int ttl) {
+        this.store = storename;
+        this.ttl = ttl;
+    }
+
+    public void handle(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        if (!"POST".equals(request.getMethod())) {
```

Comment:
Only handling POST requests --> to be discussed, maybe

Reply:
I'm not sure if you mean is should filter out all of these instead of accepting just POST?
Also javax.ws.rs.HttpMethod declares these as constants, except TRACE, so not sure the full list is already available somewhere else?

Reply:
Yes I mean filter out all these.
See what `NuxeoCorsCsrfFilter` does for the constants.